celery_pubsub/pubsub.py

```diff
@@ -1,61 +1,37 @@
+import re
 import celery
 
-__all__ = ['publish', 'publish_now', 'subscribe', 'unsubscribe', ]
-
 
 class PubSubManager(object):
+
     def __init__(self):
-        super(PubSubManager, self).__init__()
-        self.subscribed = []
         self.jobs = {}
 
     def publish(self, topic, *args, **kwargs):
-        result = self.get_jobs(topic).delay(*args, **kwargs)
-        return result
+        group = []
+        for (regex, task) in self.jobs.values():
+            if regex.match(topic):
+                group.append(task)
+        out = celery.group(group)
+        return out
```

Comment:
That is the code of the project, eventually it should rely on some kind of [radix tree](https://en.wikipedia.org/wiki/Radix_tree).

Reply:
We could think of a radix tree, why not. But the current naive implementation already caches the `celery.group` that corresponds to a given `topic`. So, this is computed once... unless `subscribe()` or `unsubscribe()` gets called, in which case it gets cleared up.

I fail to see how the radix tree lookup complexity can beat the current `O(1)` :).

`Simple is better than complex.`

Reply:
Moreover, you forgot (probably a typo) to call `.delay(*args, **kwargs)`.

Reply:
> in which case it gets cleared up.

That is exactly the point of using a radix tree, to avoid to clear (and rebuild, hence do a copy) of the primary source of thruth datastructure.  I linked the wikipedia as background reference, in the case of `celery-pubsub` the implementation as simple as a nested `dict`.

Reply:
Depends on the use case. I'd imagine people to typically (un)subscribe at the beginning of some initialisation, before publishing. Not necessarily the case, of course.
And the number of calls on publish >> number of calls on (un)subscribe.

For now, I'm still convinced that this small extra cost is worth it, as it guarantees that the publish will cost `O(1)` independently on the number of subscribed tasks in the whole system... after the first call.